gpytorch/kernels/kernel.py

```diff
@@ -517,11 +588,42 @@ def __call__(
             return res
 
         else:
-            if settings.lazily_evaluate_kernels.on():
-                res = LazyEvaluatedKernelTensor(x1_, x2_, kernel=self, last_dim_is_batch=last_dim_is_batch, **params)
+            if (settings.lazily_evaluate_kernels.on() and self._lazily_evaluate) or last_dim_is_batch:
+                num_outputs_per_input = self.num_outputs_per_input(x1_, x2_)
+                named_parameters = tuple(self.named_parameters())
+
+                if last_dim_is_batch:
+                    x1_ = x1_.transpose(-1, -2).unsqueeze(-1)
+                    x2_ = x2_.transpose(-1, -2).unsqueeze(-1)
+
+                if len(named_parameters):
+                    param_names, params = zip(*named_parameters)
+                    param_batch_shapes = [self.batch_shape] * len(params)
+                    if last_dim_is_batch:
+                        params = [
+                            param.unsqueeze(len(param_batch_shape)).transpose(-1, len(param_batch_shape))
+                            for param, param_batch_shape in zip(params, param_batch_shapes)
+                        ]
+                        param_batch_shapes = [
+                            torch.Size([*param_batch_shape, x1_.size(-3)]) for param_batch_shape in param_batch_shapes
+                        ]
+                    res = KernelLinearOperator(
+                        x1_,
+                        x2_,
+                        *params,
+                        covar_func=self._kernel_linear_operator_covar_func,
+                        num_outputs_per_input=num_outputs_per_input,
+                        param_batch_shapes=param_batch_shapes,
+                        param_names=param_names,
+                        **kwargs,
+                    )
+                else:
+                    res = KernelLinearOperator(
+                        x1_, x2_, covar_func=self.forward, num_outputs_per_input=num_outputs_per_input, **kwargs
+                    )
```

Comment:
We're pretty deep in the indentations here, might make sense to pull out some of the above into helper functions to make it clear what's going on here / easier to read the code